stdlib/src/builtin/string.mojo

```diff
@@ -2109,3 +2205,83 @@ fn _calc_format_buffer_size[type: DType]() -> Int:
         return 64 + 1
     else:
         return 128 + 1  # Add 1 for the terminator
+
+
+# ===----------------------------------------------------------------------===#
+# Format method structures
+# ===----------------------------------------------------------------------===#
+
+
+@value
+struct _FormatCurlyEntry:
+    """
+    Internally used by the `format()` method.
+
+    Purpose of the variant types:
+
+    - `String` for **kwargs
+
+        ```mojo
+        "{message}".format(message="hello")
+        ```
+
+        value field contains `message`
+
+    - `Int` for manual indexing of *args
+
+        ```mojo
+        "{0}".format("hello")
+        ```
+
+        value field contains `0`
+
+    - `NoneType` for automatic indexing of *args
+
+        ```mojo
+        "{}".format("hello")
+        ```
+
+        value field contains `None`
+
+    `*_curly` are the indexes of the enclosing curlies
+
+    This struct only supports theses:
+
+    `{number|name|nothing}`
+
+    """
+
+    var first_curly: Int
+    var last_curly: Int
+    var value: Variant[String, Int, NoneType]
+
+    @staticmethod
+    fn create_entries(arg: String) -> List[Self]:
+        """Used internally by the `format()` method.
+
+        Args:
+            arg: The "format" part provided by the user.
+
+        Returns:
+            A `List` of structured format entries.
+        """
+        var Entries = List[Self]()
+        var start = Optional[Int](None)
+        for i in range(len(arg)):
+            if arg[i] == "{":
```

Comment:
What happens with nested braces? e.g. `"{{0}}".format("foo")`? And what if it's inbalanced, e.g. `"{{0}".format("foo")`?

Reply:
:+1:  @laszlokindrat,
New commit contains more tests, including thoses two!
We can always add more if needed;
Let's keep this review conversation unresolved for now, so we remember later ?

Reply:
➡️ Updated the test to reflect what python does with theses (raising errors and escaping)

Reply:
@laszlokindrat are we good with the tests ?